File: homing-studio-base/src/main/java/hue/captains/singapura/js/homing/studio/base/composed/ListSegmentRenderer.java
package hue.captains.singapura.js.homing.studio.base.composed;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;
import hue.captains.singapura.js.homing.core.ModuleImports;
import hue.captains.singapura.js.homing.studio.base.css.StudioStyles;

import java.util.List;

/**
 * Inline renderer for {@link UnorderedListSegment} / {@link OrderedListSegment}.
 * Each item is a {@link Listable} segment dispatched — recursively, via
 * {@code ctx.renderContent} — into its own {@code <li>} on a per-item sub-branch.
 * Items are never lists (compile-time), so the recursion is depth-1.
 */
public record ListSegmentRenderer() implements DomModule<ListSegmentRenderer> {

    public static final ListSegmentRenderer INSTANCE = new ListSegmentRenderer();

    public record renderUnorderedListSegment() implements Exportable._Constant<ListSegmentRenderer> {}
    public record renderOrderedListSegment() implements Exportable._Constant<ListSegmentRenderer> {}

    @Override
    public ImportsFor<ListSegmentRenderer> imports() {
        return ImportsFor.<ListSegmentRenderer>builder()
                .add(new ModuleImports<>(List.of(
                        new StudioStyles.st_section(),
                        new StudioStyles.st_error()
                ), StudioStyles.INSTANCE))
                .build();
    }

    @Override
    public ExportsOf<ListSegmentRenderer> exports() {
        return new ExportsOf<>(INSTANCE,
                List.of(new renderUnorderedListSegment(), new renderOrderedListSegment()));
    }
}

File: homing-studio-base/src/main/resources/homing/js/hue/captains/singapura/js/homing/studio/base/composed/ListSegmentRenderer.js
// =============================================================================
// ListSegmentRenderer — ordered / unordered lists of homogeneous Listable items.
//
// renderUnorderedListSegment(branch, parent, seg, ctx) → void   (<ul>)
// renderOrderedListSegment(branch, parent, seg, ctx)   → void   (<ol>)
//
// seg shape: { anchor, items: [<segment object>, ...] }
//
// Each item is a segment, dispatched recursively through ctx.renderContent into
// its own <li> on a per-item sub-branch (so item renderers don't collide on
// element names). Items are never lists (enforced server-side), so depth is 1.
// =============================================================================

function _renderList(tag, branch, parent, seg, ctx) {
    var section = branch.createElement('section', 'section');
    css.addClass(section, st_section);
    section.id = seg.anchor;

    var list = branch.createElement('list', tag);
    var items = seg.items || [];
    for (var i = 0; i < items.length; i++) {
        var li = document.createElement('li');
        var itemBranch = branch.createBranch('item' + i);
        // ctx.renderContent activates itemBranch and dispatches the item segment.
        ctx.renderContent(items[i], li, itemBranch, seg.anchor + '-' + i);
        list.appendChild(li);
    }
    section.appendChild(list);

    parent.appendChild(section);
}

function renderUnorderedListSegment(branch, parent, seg, ctx) {
    _renderList('ul', branch, parent, seg, ctx);
}

function renderOrderedListSegment(branch, parent, seg, ctx) {
    _renderList('ol', branch, parent, seg, ctx);
}
